lib/compress/zstd_opt.c

```diff
@@ -733,6 +735,61 @@ U32 ZSTD_insertBtAndGetAllMatches (
         }
     }
 
+    /* Handle long distance matches if applicable. */
+    if (mnum > 0 && ms->ldmSeqStore && ms->ldmSeqStore->size > 0) {
+        rawSeqStore_t* ldmSeqStore = ms->ldmSeqStore;
+        int ldmIndex = ZSTD_ldm_hasMatchAtAbsolutePosition(ldmSeqStore, curr);
+        if (ldmIndex >= 0) {
+            rawSeq possibleLdm = ldmSeqStore->seq[ldmIndex];
+            U32 matchLength = possibleLdm.matchLength;
+            U32 offset = possibleLdm.offset;
+
+            /* needed for ZSTD_ldm_maybeSplitSequence() to split the correct seq */
+            ldmSeqStore->pos = ldmIndex;
+            ldmSeqStore->bytesRead = curr - 1;
+
+            /* TODO: as followup, also insert matches that are shorter than bestLength, but with smaller offset */
+            if (matchLength >= bestLength) {
+                /* TODO: This calculation needs more scrutiny, but seems to work to split matchLengths
+                 * that are longer than the remaining bytes in a block. But this doesn't work for litLengths
```

Comment:
I think it would be better to bound `matchLength` to a maximum that doesn't cross the block boundary.
This is likely possible, because `ZSTD_insertBtAndGetAllMatches()` must already do it for any regular match,
so the information about the bound must be present.